src/ffi2/datetime.rs

```diff
@@ -0,0 +1,354 @@
+    pub static mut PyDateTime_TZInfoType: PyTypeObject;
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct PyDateTime_CAPI {
+    pub DateType: *mut PyTypeObject,
+    pub DateTimeType: *mut PyTypeObject,
+    pub TimeType: *mut PyTypeObject,
+    pub DeltaType: *mut PyTypeObject,
+    pub TZInfoType: *mut PyTypeObject,
+
+    pub Date_FromDate: Option<
+        unsafe extern "C" fn(year: c_int, month: c_int, day: c_int, cls: *mut PyTypeObject)
+            -> *mut PyObject,
+    >,
+    pub DateTime_FromDateAndTime: Option<
+        unsafe extern "C" fn(
+            year: c_int,
+            month: c_int,
+            day: c_int,
+            hour: c_int,
+            minute: c_int,
+            second: c_int,
+            microsecond: c_int,
+            tzinfo: *mut PyObject,
+            cls: *mut PyTypeObject,
+        ) -> *mut PyObject,
+    >,
+    pub Time_FromTime: Option<
+        unsafe extern "C" fn(
+            hour: c_int,
+            minute: c_int,
+            second: c_int,
+            microsecond: c_int,
+            tzinfo: *mut PyObject,
+            cls: *mut PyTypeObject,
+        ) -> *mut PyObject,
+    >,
+    pub Delta_FromDelta: Option<
+        unsafe extern "C" fn(
+            days: c_int,
+            seconds: c_int,
+            microseconds: c_int,
+            normalize: c_int,
+            cls: *mut PyTypeObject,
+        ) -> *mut PyObject,
+    >,
+    pub DateTime_FromTimestamp: Option<
+        unsafe extern "C" fn(cls: *mut PyTypeObject, args: *mut PyObject, kwargs: *mut PyObject)
+            -> *mut PyObject,
+    >,
+    pub Date_FromTimestamp:
+        Option<unsafe extern "C" fn(cls: *mut PyTypeObject, args: *mut PyObject) -> *mut PyObject>,
+}
+
+// Type struct wrappers
+
+const _PyDateTime_DATE_DATASIZE: usize = 4;
+const _PyDateTime_TIME_DATASIZE: usize = 6;
+const _PyDateTime_DATETIME_DATASIZE: usize = 10;
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Date {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_DATE_DATASIZE],
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Time {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_TIME_DATASIZE],
+    pub tzinfo: *mut PyObject,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_DateTime {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_DATETIME_DATASIZE],
+    pub tzinfo: *mut PyObject,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Delta {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub days: c_int,
+    pub seconds: c_int,
+    pub microseconds: c_int,
+}
+
+// C API Capsule
+unsafe impl Sync for PyDateTime_CAPI {}
+
+lazy_static! {
+    pub static ref PyDateTimeAPI: PyDateTime_CAPI = unsafe { PyDateTime_IMPORT() };
+}
+
+#[inline(always)]
```

Comment:
Unless we have specific profiling data, I'm for doing `#[inline]` and letting the compiler decide instead of `#[inline(always)]`.

Reply:
Fine by me. I went with `#[inline(always)]` because these are macros in the C API and I figured always-inline better replicates that behavior.